### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by gauntface**
@wibblymat 

First crack at adding some docs.

Build them with: esdoc -c esdoc.json  

This uses esdoc with  "esdoc-es7-plugin" to work with the async and await stuff (which blocked us from using jsdoc).

Generally the output seems ok and it has ability to add a manual should we want it.

If this seems ok, I'm happy scrap this PR and just focus on automating deploying to github pages (or something similar) OR we can take this PR, everyone work on improving current comments and I'll still add the github pages stuff ASAP.

I've only looked at the PushClient.js to see what kind of output we'd get. Only thing that we might want to change code wise is with variables like this.endpoint, this.scope etc are all treated as public in the docs, we can make then private with comments:

    /**
     * @private
     */

Or we can make the variable names start with _ and then it's treated as private.

### CODE REVIEW

File: src/client/push-client.js [hunk 1]
```diff
@@ -65,7 +65,27 @@ let registrationReady = function(registration) {
...
+ */
 export default class PushClient {
+  /**
+   * Constructs a new PushClient.
+   *
+   * If the current browser has a push scription then it will be
+   * obtained in the constructor and sent to the endpointUrl if supplied
+   * and a subscriptionChange event will be dispatched.
+   *
+   * @param {Object} [options] - Options object should be included if you
```
Comment (wibblymat) [inline src/client/push-client.js]: I don't think you need any of the square brackets around parameter names (https://esdoc.org/tags.html#object)

File: src/client/push-client.js [hunk 2]
```diff
@@ -65,7 +65,27 @@ let registrationReady = function(registration) {
...
 
+/**
+ * PushClient is a front end library that simplifies adding push to your
+ * site.
+ */
 export default class PushClient {
+  /**
+   * Constructs a new PushClient.
+   *
+   * If the current browser has a push scription then it will be
```
Comment (wibblymat) [inline src/client/push-client.js]: scription-> subscription

File: src/client/push-client.js [hunk 3]
```diff
@@ -96,6 +116,18 @@ export default class PushClient {
...
+  /**
+   * This method will subscribe a use for push messaging.
+   *
+   * If permission isn't granted for push, this method will show the
+   * permissions dialog before attempting to subscribe the user to push.
+   *
+   * If an endpointUrl is supplied to the constructor, this will recieve
+   * a subscribe event.
+   *
+   * @return {Promise<PushSubscription>} Returns a Promise that
```
Comment (wibblymat) [inline src/client/push-client.js]: I think that the word 'Returns' in the description is redundant, here and elsewhere.

**Comment by wibblymat** [comment]
I finally got around to doing this, sorry!

**Comment by gauntface** [comment]
@wibblymat just updated this to latest code, could you give it a review and if is all good I'll take a look at generating to github pages automagically.

File: src/client/push-client.js [hunk 4]
```diff
@@ -175,24 +217,42 @@ export default class PushClient extends EventDispatch {
...
+        userId: this._userId
       });
     }
   }
 
+  /**
+   * If a service worker is registered for push, this method will return a
+   * promise that will resolve with the registration for that service worker.
+   *
+   * If there is no registration null will be the resolved reponse.
```
Comment (wibblymat) [inline src/client/push-client.js]: "If there is no registration the promise will resolve with null."

File: src/client/push-client.js [hunk 5]
```diff
@@ -175,24 +217,42 @@ export default class PushClient extends EventDispatch {
...
+   *
+   * @return {Promise<ServiceWorkerRegistration>} A Promise that
```
Comment (wibblymat) [inline src/client/push-client.js]: If the method description says what is returned then I tend to skip the description for the return type here. It avoids writing the same thing twice.

So I would tend to get for either:

```javascript
/**
 * Get the registration of the service worker being used for push
 * @return {Promise<?ServiceWorkerRegistration>} A promise that resolves to either the current
 * registration, or to null if none.
 */
```

OR

```javascript
/**
 * Returns a promise that resolves to the registration of the service worker being used for push. If
 * there is no current registration the promise resolves to null
 * @return {Promise<?ServiceWorkerRegistration>}
 */
```

**Comment by wibblymat** [comment]
Sorry it took me so long to do this :/

Basically LGTM, consider the long comment to be for future reference if you don't want to deal with it now. The other comment is probably needed though as the current sentence is sort of hard to parse.

**Comment by wibblymat** [comment]
(merge at will)
